jormungandr/src/explorer/graphql/mod.rs

```diff
@@ -162,28 +201,42 @@ impl From<blockcfg::BlockDate> for BlockDate {
     }
 }
 
+#[derive(Clone)]
 struct Transaction {
     id: FragmentId,
-    in_block: HeaderHash,
+    in_block: Option<HeaderHash>,
 }
 
 impl Transaction {
     fn from_id(id: FragmentId, context: &Context) -> FieldResult<Transaction> {
-        let in_block =
-            context
-                .db
-                .find_block_by_transaction(&id)
-                .wait()?
-                .ok_or(ErrorKind::NotFound(format!(
-                    "transaction not found: {}",
-                    &id,
-                )))?;
+        let in_block = Self::get_in_block(&id, context).ok_or(ErrorKind::NotFound(format!(
+            "transaction not found: {}",
+            &id,
+        )))?;
 
-        Ok(Transaction { id, in_block })
+        Ok(Transaction {
+            id,
+            in_block: Some(in_block),
+        })
+    }
+
+    fn from_valid_id(id: FragmentId) -> Transaction {
+        Transaction { id, in_block: None }
+    }
+
+    fn get_in_block(id: &HeaderHash, context: &Context) -> Option<HeaderHash> {
+        context.db.find_block_by_transaction(&id).wait().unwrap()
     }
 
     fn get_block(&self, context: &Context) -> FieldResult<ExplorerBlock> {
-        context.db.get_block(&self.in_block).wait()?.ok_or(
+        let block_id = match self.in_block {
+            Some(block_id) => block_id,
+            None => Self::get_in_block(&self.id, context).ok_or(ErrorKind::InternalError(
+                "Transaction's block was not found".to_owned(),
+            ))?,
+        };
+
+        context.db.get_block(&block_id).wait()?.ok_or(
```

Comment:
Why do you need a separate lookup just to check if the block is there, before obtaining it? Is the database error not descriptive enough to detect this condition?

Reply:
Agree. The `Transaction::get_in_block` non-methodic method could be named `check_hash_in_block` or something like that.